### PR TITLE
Add cancellation into response statistics

### PR DESCRIPTION
**Pull request by kthui**
Related PRs:
- https://github.com/triton-inference-server/core/pull/329
- https://github.com/triton-inference-server/common/pull/113
- https://github.com/triton-inference-server/square_backend/pull/17

Add tests for reporting cancellation statistics to response statistics.

### CODE REVIEW

File: qa/L0_response_statistics/response_statistics_test.py [hunk 1]
```diff
@@ -59,8 +60,10 @@ def callback(result, error):
 
     # Send an infer request and return its responses. 'number_of_responses' is the sum
     # of success, fail and empty responses the model should return for this request.
-    # This function waits until all success and fail responses are received.
-    def _stream_infer(self, number_of_responses):
+    # 'cancel_at_response_size' will cancel the stream when the number of responses
+    # received equals the size, set to None if cancellation is not required. This
+    # function waits until all success and fail responses are received, or cancelled.
+    def _stream_infer(self, number_of_responses, cancel_at_response_size):
```
Comment (rmccorm4) [inline qa/L0_response_statistics/response_statistics_test.py:66]: maybe just use `default = None` or `default=0` and remove the need to pass the `None` everywhere else?
```suggestion
    def _stream_infer(self, number_of_responses, cancel_at_response_size=None):
```

File: src/grpc/grpc_server.cc [hunk 2]
```diff
@@ -1186,6 +1186,20 @@ CommonHandler::RegisterModelStatistics()
...
+              GOTO_IF_ERR(err, earlyexit);
+
+              uint64_t val;
+              err = stat_json.MemberAsUInt("count", &val);
+              GOTO_IF_ERR(err, earlyexit);
+              res.mutable_cancel()->set_count(val);
+              err = stat_json.MemberAsUInt("ns", &val);
+              GOTO_IF_ERR(err, earlyexit);
+              res.mutable_cancel()->set_ns(val);
+            }
```
Comment (rmccorm4) [inline src/grpc/grpc_server.cc:1201]: optional opportunity for cleanup while you're working on this part of the code, can be a separate PR if it interests you

---

Do you think all of these verbose blocks could be reduced to something like:
```cpp
void SetLatencyCountNs(std::string stat_name, <some_type> mutable_stat) {
              triton::common::TritonJson::Value stat_json;
              err = res_json.MemberAsObject(stat_name, &stat_json);
              GOTO_IF_ERR(err, earlyexit);

              uint64_t val;
              err = stat_json.MemberAsUInt("count", &val);
              GOTO_IF_ERR(err, earlyexit);
              mutable_stat->set_count(val);
              err = stat_json.MemberAsUInt("ns", &val);
              GOTO_IF_ERR(err, earlyexit);
              mutable_stat->set_ns(val);
}

void RegisterModelStatistics() {
...
  // Re-use helper instead of all these copy/pasted blocks
  SetLatencyCountNs("compute_infer", res.mutable_compute_infer());
  SetLatencyCountNs("compute_output", res.mutable_compute_output());
  ...
  SetLatencyCountNs("cancel", res.mutable_cancel());
}

```

File: qa/L0_response_statistics/response_statistics_test.py (hunk 1 above)
Comment (kthui) [reply qa/L0_response_statistics/response_statistics_test.py]: Sure. Updated to use default arguments.

File: src/grpc/grpc_server.cc [hunk 3]
```diff
@@ -1116,167 +1030,90 @@ CommonHandler::RegisterModelStatistics()
...
+              memory_usage->set_id(id);
+            }
+            {
+              uint64_t byte_size;
+              err = usage.MemberAsUInt("byte_size", &byte_size);
+              GOTO_IF_ERR(err, earlyexit);
+              memory_usage->set_byte_size(byte_size);
+            }
           }
         }
```
Comment (kthui) [inline src/grpc/grpc_server.cc:1118]: Placed in a `{}` for better grouping, nothing else is changed.

**Review by rmccorm4 — APPROVED** [review]
Nice refactor on the stats! Sanity check that it doesn't break any existing stats or metrics related tests would be good. Great work Jacky :rocket:

File: src/grpc/grpc_server.cc (hunk 2 above)
Comment (kthui) [reply src/grpc/grpc_server.cc]: [Refactor RegisterModelStatistics()](https://github.com/triton-inference-server/server/pull/6904/commits/55ffd8eb11c2e305100eb47daa7bc4e1e73102b5)
